skipruntime-ts/adapters/kafka/src/index.ts

```diff
@@ -0,0 +1,131 @@
+ *
+ * This adapter allows to easily create Kafka _consumers_, subscribing to external event/message streams and exposing their content as eager collections within the Skip runtime.
+ *
+ * For a usage example, refer [here](https://github.com/SkipLabs/skip/tree/main/examples/chatroom).
+ *
+ * @remarks
+ * Subscription `params` may specify a `fromBeginning` boolean, which controls whether to consume all Kafka messages from the cluster, or only those sent since the creation of this external resource.
+ */
+export class KafkaExternalService implements ExternalService {
+  private consumers: Map<string, Consumer> = new Map<string, Consumer>();
+  private consumerOptions: Omit<ConsumerConfig, "groupId">;
+  private client: Kafka;
+  private messageProcessor: (msg: {
+    key: string;
+    value: string;
+    topic: string;
+  }) => Iterable<[Json, Json]>;
+
+  constructor(
+    kafka_config: {
+      clientId: string;
+      brokers: { host: string; port: number }[];
+      logLevel?: kafkaLogLevel;
+    },
+    messageProcessor: (msg: {
+      key: string;
+      value: string;
+      topic: string;
+    }) => Iterable<[Json, Json]> = (msg) => [[msg.key, msg.value]],
+    consumerOptions: Omit<ConsumerConfig, "groupId"> = {},
+  ) {
+    const brokers = kafka_config.brokers.map(
+      (obj) => `${obj.host}:${obj.port}`,
+    );
+    const logLevel: kafkaLogLevel =
+      kafka_config.logLevel ?? kafkaLogLevel.ERROR;
+
+    this.messageProcessor = messageProcessor;
+    this.consumerOptions = consumerOptions;
+    this.client = new Kafka({ ...kafka_config, brokers, logLevel });
+  }
+
+  /**
+   * Subscribe to a resource provided by the external service.
+   *
+   * @param instance - Instance identifier of the external resource.
+   * @param topic - Name of the Kafka topic to expose as a resource.
+   * @param params - Parameters of the external resource. `params.fromBeginning` controls whether to consume all Kafka messages from the cluster, or only from the creation of this external resource.
+   * @param callbacks - Callbacks to react on error/loading/update.
+   * @param callbacks.error - Error callback.
+   * @param callbacks.loading - Loading callback.
+   * @param callbacks.update - Update callback.
+   * @returns {void}
+   */
+  subscribe(
+    instance: string,
+    topic: string,
+    params: Json & { fromBeginning?: boolean },
+    callbacks: {
+      update: (updates: Entry<Json, Json>[], isInit: boolean) => void;
+      error: (error: Json) => void;
+      loading: () => void;
+    },
+  ): void {
+    callbacks.update([], true);
+    const consumer = this.client.consumer({
+      ...this.consumerOptions,
+      groupId: instance,
+    });
+    const fromBeginning = params.fromBeginning ?? true;
+    const setup = async () => {
+      await consumer.connect();
+      await consumer.subscribe({ topic, fromBeginning });
+      await consumer.run({
+        // eslint-disable-next-line @typescript-eslint/require-await
```

Comment:
This makes me slightly nervous, but I guess no way around? And I guess that the situation where this would lead to dropped messages is if the system crashes, so nothing better to do anyhow.

Reply:
I applied the same transformation (drop `async` keyword, add `return Promise.resolved(undefined)`) that you did on 795 and it lets us drop this lint suppression!